File: src/contracts/receipt.ts
/**
 * Durable Receipts — a success response means the write crossed the durable
 * boundary. "Received" alone is never a success receipt.
 */

import type { CommitIdentity, RepositoryIdentity } from "./identity.ts";

export type ReceiptDisposition = "APPLIED" | "REPLAYED";

export interface CommitReceipt {
  readonly receiptRef: string;
  readonly commandId: string;
  readonly surface: "COMMIT";
  readonly disposition: ReceiptDisposition;
  readonly result: {
    readonly repositoryId: RepositoryIdentity;
    readonly commitId: CommitIdentity;
    readonly targetRef: string;
    readonly oldCommit: CommitIdentity;
    readonly newCommit: CommitIdentity;
  };
}

export interface AppendReceipt {
  readonly receiptRef: string;
  readonly commandId: string;
  readonly surface: "APPEND";
  readonly disposition: ReceiptDisposition;
  readonly result: {
    readonly repositoryId: RepositoryIdentity;
    readonly streamRef: string;
    readonly cursor: string;
    readonly appended: readonly string[]; // record ids
  };
}

File: snapshot/treewriter/receipt.go
package treewriter

import "kc/kernel"

type ReceiptDisposition string

const (
	DispositionApplied  ReceiptDisposition = "APPLIED"
	DispositionReplayed ReceiptDisposition = "REPLAYED"
)

type Receipt struct {
	ReceiptRef  string             `json:"receiptRef"`
	CommandID   string             `json:"commandId"`
	Surface     string             `json:"surface"`
	Disposition ReceiptDisposition `json:"disposition"`
	Result      Result             `json:"result"`
}

type Result struct {
	RepositoryID kernel.RepositoryID `json:"repositoryId"`
	CommitID     kernel.CommitID     `json:"commitId"`
	TargetRef    string              `json:"targetRef"`
	OldCommit    kernel.CommitID     `json:"oldCommit"`
	NewCommit    kernel.CommitID     `json:"newCommit"`
}
